scipy/optimize/_trustregion_exact.py

```diff
@@ -0,0 +1,425 @@
+           "Numerical optimization". Springer Science &
+           Business Media. pp. 83-91.
+    .. [3] Moré, J. J., & Sorensen, D. C. (1983).
+           Computing a trust region step. SIAM Journal on Scientific
+           and Statistical Computing, 4(3), 553-572.
+    """
+
+    UPDATE_COEF = 0.01
+    EPS = np.finfo(float).eps
+
+    def __init__(self, x, fun, jac, hess, hessp=None,
+                 k_easy=0.1, k_hard=0.2):
+
+        # ``k_easy`` and ``k_hard`` are parameters used
+        # to detemine the stop criteria to the iterative
+        # subproblem solver. Take a look at pp. 194-197
+        # from reference _[1] for a more detailed description.
+
+        self.lb = None
+        self.previous_trust_radius = -1
+        self.niter = 0
+        self.k_easy = k_easy
+        self.k_hard = k_hard
+
+        # Superclass __init__
+        BaseQuadraticSubproblem.__init__(self, x, fun, jac,
+                                         hess)
+
+        # Get Lapack function for cholesky decomposition
+        self.cholesky, = get_lapack_funcs(('potrf',), (self.hess,))
+
+        # Get info about Hessian
+        self.dimension = len(self.hess)
+        self.hess_gershgorin_lb,\
+            self.hess_gershgorin_ub = gershgorin_bounds(self.hess)
+        self.hess_inf = norm(self.hess, np.Inf)
+        self.hess_fro = norm(self.hess, 'fro')
+
+        # A constant such that for vectors smaler than that
+        # backward substituition is not reliable. It was stabilished
+        # based on Golub, G. H., Van Loan, C. F. (2013).
+        # "Matrix computations". Forth Edition. JHU press., p.165.
+        self.CLOSE_TO_ZERO = self.dimension*self.EPS*self.hess_inf
+
+    def _initial_values(self, trust_radius):
+        """Given a trust radius, return a good initial guess for
+        the damping factor, the lower bound and the upper bound.
+        The values were chosen accordingly to the guidelines on
+        section 7.3.8 (p. 192) from [1].
+        """
+
+        # Upper bound
+        ub = max(0, self.jac_mag/trust_radius+min(-self.hess_gershgorin_lb,
+                                                  self.hess_fro,
+                                                  self.hess_inf))
+
+        # Lower bound
+        lb = max(0, -min(self.hess.diagonal()),
+                 self.jac_mag/trust_radius-min(self.hess_gershgorin_ub,
+                                               self.hess_fro,
+                                               self.hess_inf))
+
+        # Improve bounds with previous info
+        if trust_radius < self.previous_trust_radius:
+            lb = max(self.lb, lb)
+
+        # Damping factor
+        if lb == 0:
+            damp = 0
+        else:
+            damp = max(np.sqrt(lb*ub), lb+self.UPDATE_COEF*(ub-lb))
+
+        return damp, lb, ub
+
+    def solve(self, trust_radius):
+        """Solve quadratic subproblem"""
+
+        damp, lb, ub = self._initial_values(trust_radius)
```

Comment:
To be honest I dislike this naming trio `damp, lb, ub`. Because `damp` just sounds bad and `lb, ub` doesn't sound connected to it, so it isn't intuitive when you read the code.

How about `lambda_current, lambda_lower, lambda_upper`, or you can just rename `lambda -> alpha` (with a commentary in the code)  and go with `alpha, alpha_lower, alpha_upper`. I think the last option is the best.